File: python-algorithm/ds_3_hash/L697.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/5 11:09
# @Author  : No Name

# 使用3个哈希，1个标记元素出现频次，1个标记元素第1次出现位置，1个标记元素最后一次出现位置
class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        d1, d2, d3 ={}, {}, {}
        for i in range(len(nums)):
            if nums[i] not in d1:
                d1[nums[i]] = 1
            else:
                d1[nums[i]] += 1
            if nums[i] not in d2:
                d2[nums[i]] = i
            d3[nums[i]] = i
        min_len = len(nums)
        degree = 1
        for k,v in d1.items():
            if v>degree:
                degree = v
                min_len = d3[k]-d2[k]+1
                # 最高频词有多个的时候
            elif v==degree:
                min_len = min(min_len, d3[k]-d2[k]+1)
        return min_len

File: python-algorithm/ds_3_hash/L697.cpp

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;
class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        unordered_map<int,int> m1;
        unordered_map<int,int> m2;
        unordered_map<int,int> m3;
        for(int i=0; i<nums.size(); ++i){
            ++m1[nums[i]];
            if(m2.find(nums[i]) == m2.end())
                m2[nums[i]] = i;
            m3[nums[i]] = i;
        }
        int degree = 1;
        int len = 1;
        for(auto p:m1){
            if(p.second>degree){
                degree = p.second;
                len = m3[p.first]-m2[p.first]+1;
            }
                //都是最大度的时候，找最短的，特殊分支
            else if (p.second == degree){
                len = min(len, m3[p.first]-m2[p.first]+1);
            }
        }
        return len;
    }
};

